README.rst
=============================
django postman
=============================

.. image:: https://badge.fury.io/py/django-postman.svg
    :target: https://badge.fury.io/py/django-postman

.. image:: https://travis-ci.org/ashish2py/django-postman.svg?branch=master
    :target: https://travis-ci.org/ashish2py/django-postman

.. image:: https://codecov.io/gh/ashish2py/django-postman/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/ashish2py/django-postman

Service for email and sms


Quickstart
----------

Install django postman

    pip install git+https://github.com/ashish2py/django-postman

Add it to your `INSTALLED_APPS`:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'postman',
        ...
    )

Change your `settings.py`

.. code-block:: python

    # POSTMAIL_EMAILER
    EMAIL_BACKEND = 'postman.email_service.EmailBackend'
    
    POSTMAN_HOST = '<zaya-postman-service-hostname>'
    POSTMAN_EMAIL_ROUTE = '/api/v1/services/email/'
    POSTMAN_SMS_ROUTE = '/api/v1/services/sms/'
    POSTMAN_AUTHKEY = '<your-postman-auth-key>' 

How to use EmailBackend:

.. code-block:: python

    from django.core.mail import send_mail
    
    send_mail(
        'Subject here',
        'Here is the message.',
        'from@example.com',
        ['to@example.com'],
        fail_silently=False,
    )

How to use SMSBackend

.. code-block:: python

    from postman import sms_backend
    sms_backend.send_messages(<PHONE_NUMBER>, <MESSAGE>)
    
    # PHONE_NUMBER : accepts <country_code>
    
    Optional Parameters
    # SENDER : SMSProvider's SENDER_ID
    # ROUTER : PROMOTIONAL or TRANSACTIONAL
    
    sms_backend.send_messages(<PHONE_NUMBER>, <MESSAGE>, <SENDER_ID>, <ROUTER>)


Response Sample

.. code-block:: python

    (True, {'detail': 'success'})


Features
--------

* overrides django's send_mail and uses zaya's POSTMAN service to send email
* use POSTMAN for send messages

Credits
-------

Tools used in rendering this package:

*  Cookiecutter_
*  `cookiecutter-djangopackage`_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`cookiecutter-djangopackage`: https://github.com/pydanny/cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2017-11-28)
++++++++++++++++++

* First release on PyPI.


tests/urls.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

from django.conf.urls import url, include

from postman.urls import urlpatterns as postman_urls

urlpatterns = [
    url(r'^', include(postman_urls, namespace='postman')),
]


tests/settings.py
# -*- coding: utf-8
from __future__ import unicode_literals, absolute_import

import django

DEBUG = True
USE_TZ = True

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "j#qjj0ko&v*)k&)iqwv%!nu&rras#=&36x-_qy-t@0nkoa68m^"

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": ":memory:",
    }
}

ROOT_URLCONF = "tests.urls"

INSTALLED_APPS = [
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sites",
    "postman",
]

SITE_ID = 1

if django.VERSION >= (1, 10):
    MIDDLEWARE = ()
else:
    MIDDLEWARE_CLASSES = ()


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_django-postman
------------

Tests for `django-postman` models module.
"""

from django.test import TestCase

from postman import models


class TestPostman(TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


postman/sms_service.py
import re
import requests

from django.conf import settings
from django.core.exceptions import ValidationError

from .backends import BaseBackend


class SMSBackend(BaseBackend):
    def __init__(self, host=None, port=None, sms_route=None, auth_key=None, headers=None,
                 fail_silently=False, **kwargs):
        self.host = settings.POSTMAN_HOST
        self.route = settings.POSTMAN_SMS_ROUTE
        self.sms_route = '{host}{route}'.format(host=self.host, route=self.route)
        self.auth_key = settings.POSTMAN_AUTHKEY
        self.headers = {'application-key': self.auth_key}

    @property
    def connection(self):
        request = requests.head(self.sms_route, headers=self.headers)
        if request.status_code == 200:
            return True
        return False

    def _send_msg_to_postman(self, postman_body):
        request = requests.post(self.sms_route, json=postman_body, headers=self.headers)
        if request.status_code == 201:
            return True, request.json()
        return False, request.json()

    def validate_phone_number(self, value):
        if not re.match('\d{10}', value):
            raise ValidationError(u'%s is not valid phone number' % value)

    def send_messages(self, reciever, message, sender_id=None, route=None):

        # validate phone number
        reciever = str(reciever)
        self.validate_phone_number(reciever)

        postman_body = {
            "receiver": {
                "contact": reciever
            },
            "data": {
                "body": message
            },
            "extras": {
                "sender_id": sender_id if sender_id is not None else 'ENGDUN',
                "route": route if route is not None else 'clickhere'
            }
        }

        # send message to postman
        return self._send_msg_to_postman(postman_body)


postman/email_service.py
import requests

from django.conf import settings
from django.core.mail.backends.base import BaseEmailBackend


class EmailBackend(BaseEmailBackend):
    def __init__(self, host=None, port=None, username=None, password=None,
                 use_tls=None, fail_silently=False, use_ssl=None, timeout=None,
                 ssl_keyfile=None, ssl_certfile=None,
                 **kwargs):
        super(EmailBackend, self).__init__(fail_silently=fail_silently)
        self.host = settings.POSTMAN_HOST
        self.route = settings.POSTMAN_EMAIL_ROUTE
        self.email_route = '{host}{route}'.format(host=self.host, route=self.route)
        self.auth_key = settings.POSTMAN_AUTHKEY
        self.headers = {'application-key': self.auth_key}

    @property
    def connection(self):
        request = requests.head(self.email_route, headers=self.headers)
        if request.status_code == 200:
            return True
        return False

    def send_messages(self, email_messages):
        if not email_messages:
            return None

        if not self.connection:
            # failing if connection failed.
            return None

        for message in email_messages:
            sent = self._send(message)
            num_sent = 0
            if sent:
                num_sent += 1
        return num_sent

    def _send_email_to_postman(self, postman_body):
        request = requests.post(self.email_route, json=postman_body, headers=self.headers)
        if request.status_code == 201:
            return True, request.json()
        return False, request.json()

    def _send(self, email_message):
        ''' Helper method that does the actual email sending '''
        # check email contains recipients
        if not email_message.to:
            return False

        from_email = email_message.from_email
        recipients = email_message.to
        message = email_message.body
        subject = email_message.subject
        content_type = email_message.content_subtype

        try:
            # make request_body for postman
            for recipient in recipients:
                postman_body = {
                    "receiver": {
                        "contact": recipient
                    },
                    "sender": {
                        "contact": from_email
                    },
                    "data": {
                        "body": message,
                        "subject": subject,
                        "content_type": 'text/{ct}'.format(ct=content_type)
                    }
                }

            # PASS request to POSTMAN
            return self._send_email_to_postman(postman_body)
        except Exception as e:
            pass
